proposals/stdlib-insider-docs.md

```diff
@@ -0,0 +1,136 @@
+# Stdlib Insider Docs
+
+Owen Hilyard, Created November 17, 2024
+
+**Status**: Initial Proposal
+
+## Motivation
+
+For most languages, people who work on the standard library have the ability
+to look inside of the compiler to clarify questions they have about the
+semantics of particular operations or compiler builtins. For Mojo, that is
+not the case for many people working on the standard library. As a result,
+the exact semantics of some important compiler builtins, such as
+`lit.ownership.mark_destroyed`, are not fully known to a large number of
+people working on the standard library. For example, the fact that
+`lit.ownership.mark_destroyed` still runs the destructors of fields was a
+surprise to many at the stdlib meeting. This creates issues where Modular
+employees have to catch misuses. These language-internal dialects are, like
+Mojo itself, subject to enhancements, breaking changes, and even complete
+removal. This presents a problem since the stdlib is correctness-critical code,
+and when people who don't understand the API contract of a construct use it
+in correctness-critical code, issues are bound to happen.
+
+## Proposal
+
+In order to help address this, I propose the creation of a "stdlib insider"
+document, which contains information on MLIR operations/types, WIP features and
```

Comment:
In addition to mlir stuff, I think it would also be great to have notes on "these are projects in progress", e.g. like the changes to implicit copyability.

Reply:
I agree that having some kind of work tracking for stdlib would be useful. We have a lot of in-progress work which make decent first issues (ex: converting collections to ExplicitlyCopyable) that could be better highlighted since I think some people want to contribute and then see us arguing about allocator APIs, async executors, and linear types and think that's were the bar to contribute is. It would also enable us to track what things we want to do that are blocked or functionally blocked (essentially a rewrite) on new language features, so I'll try to merge that with the list of language feature desires by the external stdlib contributors. 

@JoeLoser Would you want something like this tracked in github projects (so you can do a one-way sync to your internal work tracking) or do you want to keep it in some markdown files? If we do markdown, I think we can use something like `stdlib/docs/insider/work_tracking` to store them. The main downside of using github is likely a bunch of front-loaded work to load everything into there and some maintenance work to keep it up to date unless there's a way to give some of us permissions to add and edit projects/tasks but nothing else in the repo. I'm open to other ideas as well.

Reply:
I'm +1 for updating our stdlib projects for both:

1. In-progress work (e.g. making things explicit by removing our incremental @implicit decorators we added recently).
2. Back-burnered work (perhaps things blocked by compiler bugs) — this would be things like SBO, implicit copyability (some of that may be in a much better state now with explicit constructor control though).

We'll also update the roadmap and make it clearer which areas we're looking for help in from contributors so we can work more effectively together on bigger projects.

As for the mechanism of how we'll track the work (GH projects or Markdown), let me think about that a bit more.  Earlier this year, we switched off of GitHub and over to Linear internally for tracking issues.  I'm not sure how well GitHub _projects_ will lend itself to that, so I'll need to look into how to best organize things internally/externally without duplicating too many things.